pkg/lib/bundlev2/generate.go

```diff
@@ -0,0 +1,144 @@
+	"strings"
+
+	"gopkg.in/yaml.v2"
+	"k8s.io/kubectl/pkg/util/slice"
+)
+
+const SynthesizedCSVFileName string = "olm.clusterserviceversion.yaml"
+
+// GenerateFunc generates a CSV for the bundle based on the resource within the bundle directory.
+// The bundleDir points to where the bundle is located, outputDir is where bundle and synthesized CSV is copy/write to.
+// If overwrite is set to true, outputDir is reloaded with the exact content of bundleDir plus a synthesized CSV
+// and inplace generation deletes the CSV if exists.
+func GenerateFunc(bundleDir, outputDir string, overwrite bool) error {
+	bundleDir, err := filepath.Abs(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	if outputDir != "" {
+		outputDir, err = filepath.Abs(outputDir)
+		if err != nil {
+			return err
+		}
+	} else {
+		// if outputDir not specified, synthesis within the bundle directory.
+		outputDir = bundleDir
+	}
+
+	bundle, err := NewBundleV2FromDirectory(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	csv, err := bundle.SynthesisedCSV()
+	if err != nil {
+		return err
+	}
+
+	if bundle.existingCSVPath != "" && !overwrite {
+		return fmt.Errorf("a ClusterServiceVersion already exist in the bundle at %s", bundle.existingCSVPath)
+	}
+
+	// Only delete (overwrite) the existing CSV in a bundle if specified to overwrite for inplace generation.
+	if bundle.existingCSVPath != "" && outputDir == bundleDir {
+		if err := os.Remove(bundle.existingCSVPath); err != nil {
+			return fmt.Errorf("error overwriting ClusterServiceVersion %s, %v", bundle.existingCSVPath, err)
+		}
+	} else if outputDir != bundleDir {
+		if overwrite {
+			// Do not expect path error since path is examined earlier.
+			os.RemoveAll(outputDir)
+		}
+		if err := CopyAllFilsFromDirectoryExcept(bundleDir, outputDir, []string{bundle.existingCSVPath}); err != nil {
+			return err
+		}
+	}
+
+	// generate CSV file to the outputDir.
+	return WriteStructToFile(csv, SynthesizedCSVFileName, outputDir)
+}
+
+// WriteStructToFile writes a struct's content to a YAML file.
+func WriteStructToFile(structContent interface{}, fileName, outputPath string) error {
+	yfile, err := yaml.Marshal(structContent)
+	if err != nil {
+		return err
+	}
+
+	if _, err = os.Stat(outputPath); os.IsNotExist(err) {
+		err = os.MkdirAll(outputPath, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+
+	filePath := path.Join(outputPath, fileName)
+	f, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		cerr := f.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	var fileLen int
+	fileLen, err = f.Write(yfile)
+	if err == nil && fileLen < len(yfile) {
+		err = io.ErrShortWrite
+	}
+
+	return err
+}
+
+// CopyAllFilsFromDirectoryExcept recursively copy files from one directory to another and truncates a file if exists.
+func CopyAllFilsFromDirectoryExcept(srcPath, dstPath string, files []string) error {
+	return filepath.Walk(srcPath, func(path string, info os.FileInfo, err error) error {
```

Comment:
Do we have a good reason to support nested directories here?

Reply:
Yes. This is a questionable design. I assume that a synthesized CSV is only useful to a bundle that has the resources that created the CSV. meaning that if the output directory and bundle directory are not the same and that user wants to overwrite, we overwrite the entire bundle specified in the ourputDir so that all resources used in creating that CSV will be present in the new bundle and only those resources. (if overwrite! )